1.13.* : importstate doesn't use configuration from previous state with ConfigDirectory

### terraform-plugin-testing version
<!---
Inspect your go.mod as below to find the version, and paste the result between
the ``` marks below.

go list -m github.com/hashicorp/terraform-plugin-testing/...

If you are not running the latest version of terraform-plugin-testing, please try
upgrading because your bug may have already been fixed.
-->

```
github.com/hashicorp/terraform-plugin-testing v1.13.1
```

The issue is not present in v1.12.0

### Relevant provider source code

<!--
Paste any Go code that you believe to be relevant to the bug
When in doubt, a minimal reproduction is best
-->
To reproduce the bug, I tested with https://github.com/hashicorp/terraform-provider-hashicups/tree/main/12-final after upgrading this module to v1.13.1, terraform-plugin-framework to v1.15.0 and terraform-plugin-go to v0.27.0

```go
func TestAccOrderResource(t *testing.T) {
	resource.Test(t, resource.TestCase{
		ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,
		Steps: []resource.TestStep{
			// Create and Read testing
			{
				ConfigDirectory: config.TestStepDirectory(),
			},
			// ImportState testing
			{
				ResourceName:      "hashicups_order.test",
				ImportState:       true,
				ImportStateVerify: true,
				// The last_updated attribute does not exist in the HashiCups
				// API, therefore there is no value for it during import.
				ImportStateVerifyIgnore: []string{"last_updated"},
			},
		},
	})
}
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.
-->

`12-final/internal/provider/testdata/TestAccOrderResource/1/main.tf`

```tf
resource "hashicups_order" "test" {
  items = [
    {
      coffee = {
        id = 1
      }
      quantity = 2
    },
  ]
}

```


### Expected Behavior
<!--
What should have happened?
-->

On the second step, the import doesn't catch an error

### Actual Behavior
<!--
What actually happened?
-->
On the second step, the module catch this error when using `ConfigDirectory: config.TestStepDirectory()` on the previous step before import : 

```bash
=== RUN   TestAccOrderResource
    order_resource_test.go:11: Step 2/2 error running import: exit status 1
        Error: resource address "hashicups_order.test" does not exist in the configuration.
        
        Before importing this resource, please create its configuration in the root module. For example:
        
        resource "hashicups_order" "test" {
          # (resource arguments)
        }
        
--- FAIL: TestAccOrderResource (0.64s)
```

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->

Launch `TestAccOrderResource` test

--- a/.changes/unreleased/BUG FIXES-20250606-110444.yaml
+++ b/.changes/unreleased/BUG FIXES-20250606-110444.yaml
@@ -1,5 +1,5 @@
 kind: BUG FIXES
-body: 'importstate: Fixed bug where prior test config is not used for `ConfigFile` or `ConfigDirectory`'
+body: 'helper/resource: Fixed bug with import state mode where prior test config is not used for `ConfigFile` or `ConfigDirectory`'
 time: 2025-06-06T11:04:44.925152-04:00
 custom:
     Issue: "516"

--- a/.changes/unreleased/BUG FIXES-20250606-110444.yaml
+++ b/.changes/unreleased/BUG FIXES-20250606-110444.yaml
@@ -1,0 +1,5 @@
+kind: BUG FIXES
+body: 'importstate: Fixed bug where prior test config is not used for `ConfigFile` or `ConfigDirectory`'
+time: 2025-06-06T11:04:44.925152-04:00
+custom:
+    Issue: "516"

--- a/helper/resource/importstate/import_block_in_config_directory_test.go
+++ b/helper/resource/importstate/import_block_in_config_directory_test.go
@@ -38,7 +38,6 @@ func TestImportBlock_InConfigDirectory(t *testing.T) {
 				ResourceName:    "examplecloud_container.test",
 				ImportState:     true,
 				ImportStateKind: r.ImportBlockWithID,
-				ConfigDirectory: config.StaticDirectory(`testdata/2`),
 			},
 		},
 	})

--- a/helper/resource/importstate/import_block_in_config_file_test.go
+++ b/helper/resource/importstate/import_block_in_config_file_test.go
@@ -38,7 +38,6 @@ func TestImportBlock_InConfigFile(t *testing.T) {
 				ResourceName:    "examplecloud_container.test",
 				ImportState:     true,
 				ImportStateKind: r.ImportBlockWithID,
-				ConfigFile:      config.StaticFile(`testdata/2/examplecloud_container.tf`),
 			},
 		},
 	})

--- a/helper/resource/importstate/import_command_with_id_test.go
+++ b/helper/resource/importstate/import_command_with_id_test.go
@@ -11,6 +11,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
 	"github.com/hashicorp/terraform-plugin-go/tftypes"
 
+	"github.com/hashicorp/terraform-plugin-testing/config"
 	r "github.com/hashicorp/terraform-plugin-testing/helper/resource"
 	"github.com/hashicorp/terraform-plugin-testing/internal/testing/testprovider"
 	"github.com/hashicorp/terraform-plugin-testing/internal/testing/testsdk/datasource"
@@ -190,6 +191,60 @@ func TestImportCommand_ImportStateVerify(t *testing.T) {
 			},
 			{
 				ResourceName:      "examplecloud_thing.test",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestImportCommand_InConfigFile(t *testing.T) {
+	t.Parallel()
+
+	r.UnitTest(t, r.TestCase{
+		TerraformVersionChecks: []tfversion.TerraformVersionCheck{
+			tfversion.SkipBelow(tfversion.Version1_0_0), // ProtoV6ProviderFactories
+		},
+		ProtoV6ProviderFactories: map[string]func() (tfprotov6.ProviderServer, error){
+			"examplecloud": providerserver.NewProviderServer(testprovider.Provider{
+				Resources: map[string]testprovider.Resource{
+					"examplecloud_container": examplecloudResource(),
+				},
+			}),
+		},
+		Steps: []r.TestStep{
+			{
+				ConfigFile: config.StaticFile(`testdata/1/examplecloud_container.tf`),
+			},
+			{
+				ResourceName:      "examplecloud_container.test",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestImportCommand_InConfigDirectory(t *testing.T) {
+	t.Parallel()
+
+	r.UnitTest(t, r.TestCase{
+		TerraformVersionChecks: []tfversion.TerraformVersionCheck{
+			tfversion.SkipBelow(tfversion.Version1_0_0), // ProtoV6ProviderFactories
+		},
+		ProtoV6ProviderFactories: map[string]func() (tfprotov6.ProviderServer, error){
+			"examplecloud": providerserver.NewProviderServer(testprovider.Provider{
+				Resources: map[string]testprovider.Resource{
+					"examplecloud_container": examplecloudResource(),
+				},
+			}),
+		},
+		Steps: []r.TestStep{
+			{
+				ConfigDirectory: config.StaticDirectory(`testdata/1`),
+			},
+			{
+				ResourceName:      "examplecloud_container.test",
 				ImportState:       true,
 				ImportStateVerify: true,
 			},

--- a/helper/resource/testing_new.go
+++ b/helper/resource/testing_new.go
@@ -131,7 +131,7 @@ func runNewTest(ctx context.Context, t testing.T, c TestCase, helper *plugintest
 
 	// use this to track last step successfully applied
 	// acts as default for import tests
-	var appliedCfg string
+	var appliedCfg teststep.Config
 	var stepNumber int
 
 	for stepIndex, step := range c.Steps {
@@ -418,7 +418,7 @@ func runNewTest(ctx context.Context, t testing.T, c TestCase, helper *plugintest
 				}
 			}
 
-			appliedCfg, err = step.mergedConfig(ctx, c, hasTerraformBlock, hasProviderBlock, helper.TerraformVersion())
+			mergedConfig, err := step.mergedConfig(ctx, c, hasTerraformBlock, hasProviderBlock, helper.TerraformVersion())
 
 			if err != nil {
 				logging.HelperResourceError(ctx,
@@ -427,6 +427,19 @@ func runNewTest(ctx context.Context, t testing.T, c TestCase, helper *plugintest
 				)
 				t.Fatalf("Error generating merged configuration: %s", err)
 			}
+
+			// Preserve the step config for future test steps to use (import state)
+			confRequest := teststep.PrepareConfigurationRequest{
+				Directory: step.ConfigDirectory,
+				File:      step.ConfigFile,
+				Raw:       mergedConfig,
+				TestStepConfigRequest: config.TestStepConfigRequest{
+					StepNumber: stepNumber,
+					TestName:   t.Name(),
+				},
+			}.Exec()
+
+			appliedCfg = teststep.Configuration(confRequest)
 
 			logging.HelperResourceDebug(ctx, "Finished TestStep")
 

--- a/helper/resource/testing_new_import_state.go
+++ b/helper/resource/testing_new_import_state.go
@@ -25,7 +25,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-testing/tfversion"
 )
 
-func testStepNewImportState(ctx context.Context, t testing.T, helper *plugintest.Helper, testCaseWorkingDir *plugintest.WorkingDir, step TestStep, cfgRaw string, providers *providerFactories, stepNumber int) error {
+func testStepNewImportState(ctx context.Context, t testing.T, helper *plugintest.Helper, testCaseWorkingDir *plugintest.WorkingDir, step TestStep, priorStepCfg teststep.Config, providers *providerFactories, stepNumber int) error {
 	t.Helper()
 
 	// step.ImportStateKind implicitly defaults to the zero-value (ImportCommandWithID) for backward compatibility
@@ -105,12 +105,6 @@ func testStepNewImportState(ctx context.Context, t testing.T, helper *plugintest
 		}
 	}
 
-	var inlineConfig string
-	if step.Config != "" {
-		inlineConfig = step.Config
-	} else {
-		inlineConfig = cfgRaw
-	}
 	testStepConfigRequest := config.TestStepConfigRequest{
 		StepNumber: stepNumber,
 		TestName:   t.Name(),
@@ -118,9 +112,20 @@ func testStepNewImportState(ctx context.Context, t testing.T, helper *plugintest
 	testStepConfig := teststep.Configuration(teststep.PrepareConfigurationRequest{
 		Directory:             step.ConfigDirectory,
 		File:                  step.ConfigFile,
-		Raw:                   inlineConfig,
+		Raw:                   step.Config,
 		TestStepConfigRequest: testStepConfigRequest,
 	}.Exec())
+
+	// If the current import state test step doesn't have configuration, use the prior test step config
+	if testStepConfig == nil {
+		logging.HelperResourceTrace(ctx, "Using prior TestStep Config for import")
+
+		testStepConfig = priorStepCfg
+	}
+
+	if testStepConfig == nil {
+		t.Fatal("Cannot import state with no specified config")
+	}
 
 	switch {
 	case step.ImportStateConfigExact:
@@ -131,10 +136,6 @@ func testStepNewImportState(ctx context.Context, t testing.T, helper *plugintest
 
 	case kind.plannable():
 		testStepConfig = appendImportBlock(testStepConfig, resourceName, importId)
-	}
-
-	if testStepConfig == nil {
-		t.Fatal("Cannot import state with no specified config")
 	}
 
 	var workingDir *plugintest.WorkingDir

--- a/helper/resource/testing_new_import_state.go
+++ b/helper/resource/testing_new_import_state.go
@@ -118,13 +118,13 @@ func testStepNewImportState(ctx context.Context, t testing.T, helper *plugintest
 
 	// If the current import state test step doesn't have configuration, use the prior test step config
 	if testStepConfig == nil {
+		if priorStepCfg == nil {
+			t.Fatal("Cannot import state with no specified config")
+		}
+
 		logging.HelperResourceTrace(ctx, "Using prior TestStep Config for import")
 
 		testStepConfig = priorStepCfg
-	}
-
-	if testStepConfig == nil {
-		t.Fatal("Cannot import state with no specified config")
 	}
 
 	switch {

